cmd/snap-gdb-shim/snap-gdb-shim.c

```diff
@@ -29,6 +30,30 @@ int main(int argc, char **argv)
 			printf("-%s-\n", argv[i]);
 		}
 	}
+
+        // check if we run as SUDO and if so switch to the real gid/uid
+	const char *sudo_gid_env = getenv("SUDO_GID");
+	if (getgid() == 0 && sudo_gid_env != NULL) {
+		int sudo_gid = sc_must_parse_int(sudo_gid_env);
+		if (sudo_gid != 0) {
+			if (setgid(sudo_gid) < 0) {
+				die("cannot switch to gid %d", sudo_gid);
+			}
+		}
+	}
+	const char *sudo_uid_env = getenv("SUDO_UID");
+	if (getuid() == 0 && sudo_uid_env != NULL) {
+		int sudo_uid = sc_must_parse_int(sudo_uid_env);
+		if (sudo_uid != 0) {
+			if (setuid(sudo_uid) < 0) {
+				die("cannot switch to uid %d", sudo_uid);
+			}
+		}
+	}
+	// Ideally we would also call setgroups() now but seccomp will
+	// prevent this. At this point we are inside the confinement
+	// of the snap already.
```

Comment:
This comment should be above the setgid() code cause setgroups() cannot be run after dropping privs. Note that https://github.com/snapcore/snapd/pull/6681 introduces allowing `setgroups(0, NULL)` so perhaps add to your comment you can do this or remove the comment, add that change to this PR and do `setgroups(0, NULL)` here.

Reply:
Good point, I updated the comment. However while looking at this I noticed this PR apparently has a fatal flaw - the setuid/setgid calls happen inside the confinement (they need to run after snap-confine did its job) but at this point they get denied. I added a spread test and indeed:
```
[  561.127570] audit: type=1400 audit(1554797174.738:44): apparmor="DENIED" operation="capable" profile="snap.test-snapd-tools.env" pid=17870 comm="snap-gdb-shim" capability=6  capname="setgid"
```
This seems to be tricky to fix :/

Reply:
In https://github.com/snapcore/snapd/pull/6681 I found myself in the position of either adding implied seccomp rules that allowed the drop or to reorganize the launcher slightly to not need any implied rules. I opted for the latter.

I mention this because you seem faced with a similar dilemma: either add some capability rules the apparmor profile on the fly (I *highly* dislike this) or to rethink the invocation so that we've already dropped by the time we exec (or to somehow support --gdb without sudo). I don't have specification recommendations on this, but recommend the rethinking route.